[libpas] Support MallocStackLogging

Constellation opened the pull request:
#### 5d192fe3d632fdaec9371af3f54db375cd9afe95
<pre>
[libpas] Support MallocStackLogging
<a href="https://bugs.webkit.org/show_bug.cgi?id=246581">https://bugs.webkit.org/show_bug.cgi?id=246581</a>
rdar://101210195

Reviewed by Darin Adler.

This patch adds MallocStackLogging (MSL) support to libpas. After this change, bmalloc will not fallback to system malloc with
MallocStackLogging environment variable (Malloc environment variable etc. fallbacks), and enable libpas MSL.
The limitation is that libpas is not supporting on-demand enablement of MSL. If we would like to collect data,
MallocStackLogging environment variable needs to be set (or use system malloc as before, right now).

libmalloc is invoking malloc_logger hook function on allocations and deallocations. The other framework implements MSL&apos;s actual
stacktrace capturing and it configures malloc_logger hook function, and this is how MSL is implemented in libmalloc. We do the
same thing in libpas: calling malloc_logger hook function on allocations and deallocations.

The challenge is that how to avoid adding these code into the hottest fast path in libpas: libpas allocation path is super tightly
optimized and any additional load etc. regress performance severely. We took similar approach to our system malloc enablement in
libpas: when libpas found MallocStackLogging, it disables thread-local-cache completely. This enforces all allocations and deallocations
go to the slow path. Then, we inserted this hook in the slow path so that we insert this logging without performance penalty when
MSL is not enabled.

We ensured that MSL works and no performance penalty is observed from A/B tests.

* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/Environment.cpp:
(bmalloc::isMallocEnvironmentVariableImplyingSystemMallocSet):
* Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h:
* Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
(pas_try_deallocate_slow_no_cache):
* Source/bmalloc/libpas/src/libpas/pas_deallocate.h:
(pas_try_deallocate_not_small_exclusive_segregated):
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.c: Copied from Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h.
(compute_msl_status):
(pas_compute_msl_is_enabled):
(pas_msl_malloc_logging_slow):
(pas_msl_free_logging_slow):
* Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.h: Added.
(pas_msl_is_enabled):
(pas_msl_malloc_logging):
(pas_msl_free_logging):
* Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
(pas_thread_local_cache_can_set):
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
(pas_try_allocate_common_impl_slow):
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
(pas_try_reallocate_table_bitfit_case):
(pas_try_reallocate):

Canonical link: <a href="https://commits.webkit.org/255601@main">https://commits.webkit.org/255601@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/63ec2975db9d6baeb8cd3aa903e0ab0b77227a7a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93056 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2257 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/23652 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/102765 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162990 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2267 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30585 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85439 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98881 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/98721 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79530 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/28475 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/83474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71587 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84377 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36979 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/79444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/34791 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/79444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/38663 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/40896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/82077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1788 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40590 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/37510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/37/builds/82077 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.c:
```diff
@@ -0,0 +1,75 @@
+pas_msl_is_enabled_flag pas_msl_is_enabled_flag_value = pas_msl_is_enabled_flag_indeterminate;
+static void compute_msl_status(void)
+{
+    if (!!getenv("MallocStackLogging"))
```
I would have written this with the ternary operator to make it clear both sides of the if statement are setting the same flag.

Constellation replied:
Sounds good, changed.

darinadler commented on Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.c:
```diff
@@ -0,0 +1,75 @@
+
+PAS_BEGIN_EXTERN_C;
+
+pas_msl_is_enabled_flag pas_msl_is_enabled_flag_value = pas_msl_is_enabled_flag_indeterminate;
```
Can’t this simply be a `bool`? No code needs to read it before `compute_msl_status` status is called, so doesn’t seem like we need an indeterminate state.

Constellation replied:
We would like to avoid calling into `pas_compute_msl_is_enabled` function if possible, and reading this from inlined function can avoid it in almost all cases.

darinadler replied:
Is that double checked locking? Is it safe? https://en.wikipedia.org/wiki/Double-checked_locking

Constellation replied:
@darinadler yes, this is safe, since pthread_once emits memory barrier. And we are always going to this function call if it is indeterminate

darinadler reviewed (approved):
Looks good. Hope you did some testing; can’t tell from the pull request.

Constellation committed:
[libpas] Support MallocStackLogging
https://bugs.webkit.org/show_bug.cgi?id=246581
rdar://101210195

Reviewed by Darin Adler.

This patch adds MallocStackLogging (MSL) support to libpas. After this change, bmalloc will not fallback to system malloc with
MallocStackLogging environment variable (Malloc environment variable etc. fallbacks), and enable libpas MSL.
The limitation is that libpas is not supporting on-demand enablement of MSL. If we would like to collect data,
MallocStackLogging environment variable needs to be set (or use system malloc as before, right now).

libmalloc is invoking malloc_logger hook function on allocations and deallocations. The other framework implements MSL's actual
stacktrace capturing and it configures malloc_logger hook function, and this is how MSL is implemented in libmalloc. We do the
same thing in libpas: calling malloc_logger hook function on allocations and deallocations.

The challenge is that how to avoid adding these code into the hottest fast path in libpas: libpas allocation path is super tightly
optimized and any additional load etc. regress performance severely. We took similar approach to our system malloc enablement in
libpas: when libpas found MallocStackLogging, it disables thread-local-cache completely. This enforces all allocations and deallocations
go to the slow path. Then, we inserted this hook in the slow path so that we insert this logging without performance penalty when
MSL is not enabled.

We ensured that MSL works and no performance penalty is observed from A/B tests.

* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/bmalloc/Environment.cpp:
(bmalloc::isMallocEnvironmentVariableImplyingSystemMallocSet):
* Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h:
* Source/bmalloc/libpas/src/libpas/pas_deallocate.c:
(pas_try_deallocate_slow_no_cache):
* Source/bmalloc/libpas/src/libpas/pas_deallocate.h:
(pas_try_deallocate_not_small_exclusive_segregated):
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
* Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.c: Copied from Source/bmalloc/libpas/src/libpas/pas_darwin_spi.h.
(compute_msl_status):
(pas_compute_msl_is_enabled):
(pas_msl_malloc_logging_slow):
(pas_msl_free_logging_slow):
* Source/bmalloc/libpas/src/libpas/pas_malloc_stack_logging.h: Added.
(pas_msl_is_enabled):
(pas_msl_malloc_logging):
(pas_msl_free_logging):
* Source/bmalloc/libpas/src/libpas/pas_thread_local_cache.h:
(pas_thread_local_cache_can_set):
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
(pas_try_allocate_common_impl_slow):
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
(pas_try_reallocate_table_bitfit_case):
(pas_try_reallocate):

Canonical link: https://commits.webkit.org/255601@main
